src/bucket/BucketIndex.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "bucket/LedgerCmp.h"
+#include "lib/bloom_filter.hpp"
+#include "util/NonCopyable.h"
+#include <filesystem>
+#include <memory>
+#include <optional>
+#include <variant>
+
+namespace stellar
+{
+
+/**
+ * BucketIndex is an in-memory mapping of LedgerKey's to the file offset
+ * of the associated LedgerEntry in a given Bucket file. Because the set of
+ * LedgerKeys is too large to keep in memory, BucketIndex can either index
+ * individual keys or key ranges.
+ *
+ * For small buckets, an individual index is used for faster index lookup. For
+ * larger buckets, the range index is used. The range index cannot give an exact
+ * position for a given LedgerEntry or tell if it exists in the bucket, but can
+ * give an offset range for where the entry would be if it exists. Config flags
+ * determine the size of the range index, as well as what bucket size should use
+ * the individual index vs range index.
+ */
+
+class Config;
+
+struct BucketIndexRangeEntry
+{
+    LedgerKey lowerBound;
+    LedgerKey upperBound;
+
+    BucketIndexRangeEntry(LedgerKey low, LedgerKey high)
+        : lowerBound(low), upperBound(high)
```

Comment:
I'll definitely add the bounds check, but it would be difficult to make the index keys const. C++17 containers don't support `const` template types so I think it would be too involved to make the keys const. The index itself is a protected member and all the functions are `const` so I think we'll be safe without `const` keys.

Reply:
ok. it looks like the bounds check is still missing though

Reply:
^ looks like this comment got lost

Reply:
I believe I addressed the bounds check with `                  releaseAssert(rangeEntry.upperBound < key);` in the constructor for index `BucketIndexImpl`. This just checks to make sure the bounds are correct when constructing the range index.

Reply:
oh I thought you were going to enforce this in the constructor/at assignment in `RangeEntry` itself (in case there are other call sites now or in the future). This is not critical for landing though, and can be cleaned up later, so we can consider this comment resolved